artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ActiveMQThreadPoolExecutor.java

```diff
@@ -29,125 +30,58 @@
  */
 public class ActiveMQThreadPoolExecutor extends ThreadPoolExecutor {
 
-   @SuppressWarnings("serial")
-   private static class ThreadPoolQueue extends LinkedBlockingQueue<Runnable> {
+   /**
+    * The default rejected execution handler
+    */
+   private static final RejectedExecutionHandler defaultHandler = new AbortPolicy();
 
-      private ActiveMQThreadPoolExecutor executor = null;
+   // Handler executed when a task is submitted and a new thread cannot be created (because maxSize was reached)
+   // It queues the task on the executors's queue (using the add() method, see ThreadPoolQueue class below)
+   private static class QueueExecutionHandler implements RejectedExecutionHandler {
 
-      // keep track of the difference between the number of idle threads and
-      // the number of queued tasks. If the delta is > 0, we have more
-      // idle threads than queued tasks and can add more tasks into the queue.
-      // The delta is incremented if a thread becomes idle or if a task is taken from the queue.
-      // The delta is decremented if a thread leaves idle state or if a task is added to the queue.
-      private static final AtomicIntegerFieldUpdater<ThreadPoolQueue> DELTA_UPDATER = AtomicIntegerFieldUpdater.newUpdater(ThreadPoolQueue.class, "threadTaskDelta");
-      private volatile int threadTaskDelta = 0;
+      private final RejectedExecutionHandler handler;
 
-      public void setExecutor(ActiveMQThreadPoolExecutor executor) {
-         this.executor = executor;
+      private QueueExecutionHandler(RejectedExecutionHandler handler) {
+         Objects.requireNonNull(handler);
+         this.handler = handler;
       }
 
       @Override
-      public boolean offer(Runnable runnable) {
-         boolean retval = false;
-
-         if (threadTaskDelta > 0 || (executor.getPoolSize() >= executor.getMaximumPoolSize())) {
-            // A new task will be added to the queue if the maximum number of threads has been reached
-            // or if the delta is > 0, which means that there are enough idle threads.
-
-            retval = super.offer(runnable);
-
-            // Only decrement the delta if the task has actually been added to the queue
-            if (retval)
-               DELTA_UPDATER.decrementAndGet(this);
+      public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+         if (executor.isShutdown() || !executor.getQueue().add(r)) {
+            handler.rejectedExecution(r, executor);
          }
-
-         return retval;
       }
+   }
 
-      @Override
-      public Runnable take() throws InterruptedException {
-         // Increment the delta as a thread becomes idle
-         // by waiting for a task to take from the queue
-         DELTA_UPDATER.incrementAndGet(this);
-
-
-         Runnable runnable = null;
+   // A specialized LinkedBlockingQueue that takes new elements by calling add() but not offer()
```

Comment:
Need to change this, sorry in my example i was doing quickly, so comments etc left unchanged.

Maybe

----------------------

A specialized LinkedTransderQueue that takes new elements on offer, by trying to transfer if if idle thread by calling tryTransfer(), calling add() will force the queue to take a new element.

This is to force the ThreadPoolExecutor to offer successfully when an idle thread exists, but otherwise create new threads and never queue.
----------------------

Reply:
I don't know bud, just looking at the doc of `tryTransfert` is very clear (for a concurrent q doc obviously :P).
But If you feel that adding some doc could help I will add some for sure :+1:

Reply:
Well then we remove the current doc at least. The doc atm != the code

Reply:
True: just a qq...have you tried this new impl too?
I've just run a simple load and it seems the best of the 3, but you know...I'm not quite trusty of my box :)

Reply:
I did but i didnt capture results. It was a bit better than your commit 1, for the use case i was checking for. If you need results, ill have to try do later in the week, busy week again.

Reply:
@michaelandrepearce No worries Michael: I can ask @clebertsuconic super-charged machine as well to run a test with it :P
The primary concerns/motivation of this PR are:

- better scalability (ReentrantLock of LinkedBlockingQueue makes me super-sad!!) without sacrificing low number, but fast pairs connected
- reduction of code complexity

Reply:
@franz1981 

test is run 20 times without restart of the process, and the provided times are us average for each run of 100,000, was tested on a stock dell r630, with persistance disabled to eliminate disk io from impacting results.

commit 1)

175
165
66
65
62
60
98
98
45
45
50
50
45
52
44
44
46
47
44

latest version)

165
161
69
70
52
51
104
47
45
45
53
53
51
44
42
42
43
44
45

I think this looks good to merge.